devito/types/grid.py

```diff
@@ -774,34 +763,38 @@ def __subdomain_finalize__(self, grid, counter=0, **kwargs):
             self._local_bounds = self._global_bounds
 
         # Sanity check
-        if len(self._local_bounds) != 2*len(self.dimensions):
+        if len(self._local_bounds) != 2*len(grid.dimensions):
             raise ValueError("Left and right bounds must be supplied for each dimension")
 
         # Associate the `_local_bounds` to suitable symbolic objects that the
         # compiler can use to generate code
         n = counter - npresets
         assert n >= 0
-        self._implicit_dimension = i_dim = Dimension(name='n%d' % n)
-        functions = []
-        for j in range(len(self._local_bounds)):
-            index = floor(j/2)
-            d = self.dimensions[index]
-            if j % 2 == 0:
-                fname = "%s_%s" % (self.name, d.min_name)
-            else:
-                fname = "%s_%s" % (self.name, d.max_name)
-            f = Function(name=fname, grid=self._grid, shape=(self._n_domains,),
-                         dimensions=(i_dim,), dtype=np.int32)
 
-            # Check if shorthand notation has been provided:
-            if isinstance(self._local_bounds[j], int):
-                f.data[:] = np.full((self._n_domains,), self._local_bounds[j],
-                                    dtype=np.int32)
-            else:
-                f.data[:] = self._local_bounds[j]
+        i_dim = Dimension(name='n%d' % n)
+
+        dimensions = []
+        for i, d in enumerate(grid.dimensions):
+            dname = '%si%d' % (d.name, counter)
+            fnames = ["%s_%s_%s" % (self.name, dname, mM)
+                      for mM in ('m', 'M')]
+            f = tuple(Function(name=fname, grid=self._grid, shape=(self._n_domains,),
+                               dimensions=(i_dim,), dtype=np.int32) for fname in fnames)
 
-            functions.append(f)
-        self._functions = as_tuple(functions)
+            # Check if shorthand notation has been provided:
+            for j in range(2):
```

Comment:
why 2? looks hardcoded, wasn;t it better

Reply:
There's only ever going to be a minimum and maximum bound. This is actually enforced higher up in this function